Add test and build workflows

p-galligan opened the pull request:
Adds initial test and build/deploy workflows.

p-galligan committed:
update workflows

Adds initial outline for CI/CD workflows

p-galligan committed:
update test workflow

Updates test workflow action to run on PR against dev.

p-galligan committed:
update test workflow

Allows manually running of test workflow

p-galligan committed:
fix workflows

fix test workflow action

p-galligan committed:
update build workflow

Updates build workflow to not use deploy scripts repo

p-galligan committed:
allow deploy to run once

Add on PR for deploy for testing

p-galligan committed:
update build workflow

specify dockerfile location in build

p-galligan committed:
update build workflow

Test build workflow

p-galligan committed:
update build workflow

It helps to actually check out the repository...

p-galligan committed:
update build workflow

Uses AWS ecr login action

p-galligan committed:
update build workflow

updates date/time setting and specify account id for ECR login

p-galligan committed:
update build file

Updates build workflow for testing

p-galligan committed:
update build

testing build workflow syntax

p-galligan committed:
update build workflow

update docker push syntax

p-galligan committed:
update build workflow

more testing

p-galligan committed:
update build workflow

try different authentication method

p-galligan committed:
update build

remove no-include-email

p-galligan committed:
update build workflow

testing without password masking

p-galligan committed:
update build workflow

add quotations

p-galligan committed:
update build workflow

test against dev ECR

p-galligan committed:
update workflow

Adding both accounts to login to ecr and switching back to prod repo for push

p-galligan committed:
fix variable name

fix variable name in build workflow

p-galligan committed:
update build

remove dev account id

p-galligan committed:
remove quotes

remove quotations from ecr login

p-galligan committed:
update build

attempt different ECR login method

p-galligan committed:
revert build update

reverts ecr login method

p-galligan committed:
update build

test build variables

p-galligan committed:
Revert "update build"

This reverts commit f2d1d4deea7de0ab1dfd84b72090e50efdee8049.

p-galligan committed:
update build

test login and pull

p-galligan committed:
fix syntax

fixes workflow syntax

p-galligan committed:
test list

tests listing images

p-galligan committed:
add registry id

add registry id to test

p-galligan committed:
test push

test build and push

p-galligan committed:
use secrets correctly

change syntax

p-galligan committed:
tweaks

updates

p-galligan committed:
fix typo

get outta here

p-galligan committed:
back again

actually there were two typos

p-galligan committed:
head into desk

desk up to head?

p-galligan committed:
update workflows

come at me bro

p-galligan committed:
update with fix

Test full build/tag workflow

p-galligan committed:
fix versioned tag

fixes tag for versioned info

p-galligan committed:
fix typo

fixes typos

p-galligan committed:
remove colons and dashes

Removes colons and dashes from dates

p-galligan committed:
test prod step

tests prod step in workflows

p-galligan committed:
view manifest

See the manifest we've created

p-galligan committed:
testing manifest

test image manifest creation

p-galligan committed:
fix typo

fix manifest typo

p-galligan committed:
testing

feels so close

p-galligan committed:
remove period

removes a pesky period

p-galligan committed:
testing manifest again

tests

p-galligan committed:
fix typo

fixes another typo

p-galligan committed:
testing output

test batch image output

p-galligan committed:
test variable assign

testing

p-galligan committed:
Pass manifest

Passes manifest in prod workflow

p-galligan committed:
set env variable

Sets env variable for image manifest

p-galligan committed:
trying this

good luck

p-galligan committed:
fix typo

neverending stoooooory

p-galligan committed:
change output

change json output

p-galligan committed:
again

another round

p-galligan committed:
didn't work

that didn't go as planned

p-galligan committed:
one more

once again into the breach

p-galligan committed:
fix case

Uppercase is important

p-galligan committed:
trying again

let's see if this worked

p-galligan committed:
reinstate if statement

Uncomments the if statement for prod tagging

p-galligan committed:
Update build

Only build on push to dev and base

p-galligan committed:
remove unused variable

removes unused target image variable

p-galligan committed:
remove travis file

Removes unused travis file

helrond commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,63 @@
+          aws-region: ${{ secrets.AWS_REGION }}
+
+      - name: ECR login
+        run: aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${{ secrets.AWS_ACCOUNT_ID }}.dkr.ecr.us-east-1.amazonaws.com
```
It looks like we already have a `${{ secrets.AWS_REGION }}` defined here so let's just use it to keep things consistent.

helrond commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,63 @@
+          docker tag ${{ env.APP_NAME }} "${{ secrets.AWS_ACCOUNT_ID }}".dkr.ecr."${{ secrets.AWS_REGION }}".amazonaws.com/"${{ env.APP_NAME }}":latest
+          docker push "${{ secrets.AWS_ACCOUNT_ID }}".dkr.ecr."${{ secrets.AWS_REGION }}".amazonaws.com/"${{ env.APP_NAME }}":latest
+
+      - name: Tag and push dev to ECR
```
I don't see a conditional here which would prevent this from executing on a push to `base`.

helrond commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: Build Docker image
+        run: docker build -t ${{ env.APP_NAME }} --target build .
+
+      - name: Tag and push latest to ECR
```
I think there's an open question here about whether we want the `latest` to be created for builds in both `development` and `base` or just `base`. Because right now it looks like it's both and I'm not sure about that. Can you look at best practices around that?

helrond replied:
Also, all of these tags and pushes might be a good argument for us building our own hook that does this. Maintaining this in multiple apps feels like a bear.

p-galligan replied:
Looks like best practice is to only use latest for testing purposes. According to AWS documentation [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-considerations.html#:~:text=We%20recommend%20that%20you%20use,container%20image%20for%20every%20commit.).

> Container images are stored in a container registry. Each image in a registry is identified by a tag. There's a tag called latest. This tag functions as a pointer to the latest version of the application container image, similar to the HEAD in a git repository. We recommend that you use the latest tag only for testing purposes. As a best practice, tag container images with a unique tag for each build. We recommend that you tag your images using the git SHA for the git commit that was used to build the image.

p-galligan replied:
I'm going to update to include an update to latest only on the base branch

helrond replied:
OK let's talk some more about this. I'm wondering if we should be building at all in base, or if we should just be retagging.

helrond commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,63 @@
+          docker tag ${{ env.APP_NAME }} "${{ secrets.AWS_ACCOUNT_ID }}".dkr.ecr."${{ secrets.AWS_REGION }}".amazonaws.com/"${{ env.APP_NAME }}":dev
+          docker push "${{ secrets.AWS_ACCOUNT_ID }}".dkr.ecr."${{ secrets.AWS_REGION }}".amazonaws.com/"${{ env.APP_NAME }}":dev
+
+      - name: Tag and push versioned image to ECR
```
Again, do we want this to run on `development` too?

p-galligan replied:
I've added a conditional to only run on base.

helrond commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,63 @@
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          role-to-assume: ${{ secrets.AWS_DEPLOY_ROLE }}
+          role-skip-session-tagging: true
+          role-duration-seconds: 1000
```
15 minutes is 900 seconds - I'd be inclined to use that as our default (and documented) number.

helrond reviewed (changes requested):
Got a couple of questions and suggestions on this one.

p-galligan committed:
Updates build workflow

Updates the build workflow to more restrict when images are created and to use secrets when available.

p-galligan committed:
update build workflow

Updates build workflow to change environment.

p-galligan committed:
Update test workflow

Updates test workflow to use repository variable for environment.

p-galligan committed:
fix typos in deploy workflow

Fixes typos in conditionals for deploy workflow steps.
